torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1392,19 +1392,28 @@ void InitXlaModuleBindings(py::module m) {
                                  bool replicated = false, bool manual = false) {
     xla::OpSharding sharding =
         ShardingUtil::CreateOpSharding(tile_assignment, replicated, manual);
-
-    // TODO(yeounoh) use `SPMD` device to avoid input at::Tensor uploading.
-    // For now, we move the pre-uploaded data back to cpu_tensor for sharding.
     XLATensorPtr xtensor = bridge::GetXlaTensor(input);
-    std::vector<XLATensorPtr> xla_tensors{xtensor};
-    at::Tensor cpu_tensor = XLATensor::GetTensors(&xla_tensors)[0];
 
     // Existing annotation must be cleared explicitly. We do not clear and
     // overwrite the existing sharding on user's behalf, since it could lead to
     // confusion/error.
     XLA_CHECK(xtensor->sharding_spec() == nullptr)
         << "Existing annotation must be cleared first.";
 
+    at::Tensor cpu_tensor;
+    if (xla::sys_util::GetEnvBool("XLA_USE_SPMD", false)) {
+      // When virtual device is enabled, we expect tensor data to live on the
+      // CPU when it's passed into mark_sharding.
+      XLA_CHECK(xtensor->CurrentTensorData().has_value())
+          << "Cannot shard tensor. Data missing from CPU.";
```

Comment:
A bit more details, "When virtual device is enabled for SPMD, we defer the initial data transfer to the device and retain the original data on the host, until the sharded data transfer."

Reply:
Added